### PR TITLE
Bug 1887863: Patch Flavor Not Found validation for OpenStack Install Config

### PR DESCRIPTION
**Pull request by iamemilio**
Fixes an issue where we expected cloudInfo to not have a
key for the flavorName if it was not found. In the code and the unit tests, cloudInfo's flavor map now has a pointer value
instead of a struct. This allows us to set it to nil when the flavor is not found in openstack.

/label platform/openstack

### CODE REVIEW

File: pkg/asset/installconfig/openstack/validation/platform_test.go
Comment (Fedosin) [inline pkg/asset/installconfig/openstack/validation/platform_test.go:124]: the option is called `computeFlavor`: https://github.com/openshift/installer/blob/master/data/data/install.openshift.io_installconfigs.yaml#L1013

No need to rename it.

File: pkg/asset/installconfig/openstack/validation/machinepool_test.go
Comment (Fedosin) [inline pkg/asset/installconfig/openstack/validation/machinepool_test.go:146]: it is called "type", not "flavorName": https://github.com/openshift/installer/blob/master/data/data/install.openshift.io_installconfigs.yaml#L428

let's keep it as it was before

File: pkg/asset/installconfig/openstack/validation/machinepool.go
Comment (Fedosin) [inline pkg/asset/installconfig/openstack/validation/machinepool.go:45]: fldPath.Child("type")

**Review by Fedosin — CHANGES_REQUESTED** [review]

File: pkg/asset/installconfig/openstack/validation/machinepool.go
Comment (adduarte) [inline pkg/asset/installconfig/openstack/validation/machinepool.go:105]: could this possibly hide a problem?  The flavorName being passed int should never be empty. 
But here if it is, it is silently overlooked.  Maybe fill the errorList? or let it fall through 
The next lookup "ci.Flavors[flavorName]  will catch it

Comment (iamemilio) [reply pkg/asset/installconfig/openstack/validation/machinepool.go:105]: That should be handled by the static usage validations. There are cases where it is valid for flavor name to be empty

File: pkg/asset/installconfig/openstack/validation/platform_test.go
Comment (iamemilio) [reply pkg/asset/installconfig/openstack/validation/platform_test.go]: ack to all name changes

File: pkg/asset/installconfig/openstack/validation/machinepool.go
Comment (iamemilio) [reply pkg/asset/installconfig/openstack/validation/machinepool.go:105]: The purpose of this block of code is just to make sure that if a flavor is provided, that flavor is valid

**Comment by pierreprinetti** [comment]
> The error message was reporting errors on field "type" instead of "flavorName"

My patch #4269 changes the error message to mention "type" (and not "flavorName"), because [that seems to be the name of the JSON property in MachinePool](https://github.com/openshift/installer/blob/31dba4362530c3afaacf20e3a7971a10c1d9f288/pkg/types/openstack/machinepool.go#L8).

> cloudInfo's flavor map now has a pointer value instead of a struct

Can you help me understand why a pointer is necessary? The "comma OK" idiom is how you would check if a value was in the map. See this example: https://play.golang.org/p/Ubj2NqhG9pg

---

I am not sure what case you want to address with this patch; perhaps you could create a bug in Bugzilla that describes what happens vs what you expected to happen?

/hold

**Comment by iamemilio** [comment]
https://issues.redhat.com/browse/OSASINFRA-2133

**Comment by iamemilio** [comment]
The pointer lets us set the entire value to nil in the Flavors map when openstack fails to find a given flavor, rather than setting it to Flavor{}. This lets us more easily indicate that it was not found. The way the library is built, it is not optimal for us to exclude lookups that failed from the Flavors map, since we use it as an optomization to check which openstack queries we have already run to avoid duplicating calls to the openstack api.

**Comment by iamemilio** [comment]
NOTE: needs squash before merge

**Comment by pierreprinetti** [comment]
I think that the problem is here:
https://github.com/openshift/installer/blob/ac6f233c352843294abf9719fe5659527cdc5bf6/pkg/asset/installconfig/openstack/validation/cloudinfo.go#L175-L178

If you just remove these 3 lines, the validation errors with a nice message when the flavor is not found by name.

@Fedosin Do you perhaps remember why we are silently ignoring a FlavorNotFound error?

**Comment by iamemilio** [comment]
We try to defer errors caused by user error in the install config to a section of code that can append it to the error list so that the customer gets all of their install config validation errors in a list.

**Comment by Fedosin** [comment]
@pierreprinetti during cloud info generation we check only critical errors that cause the installer to stop immediately (auth problems, 5xx errors). Here we do not consider ErrResourceNotFound critical to allow the validator to check other install config fields later, and return a full list of issues.

**Comment by iamemilio** [comment]
/assign @Fedosin 

Going to assign mike too since he is knowledgeable on this code

**Comment by pierreprinetti** [comment]
/assign

Turning a map of structs into a map of pointers to pass the message that the flavor might not have been found is not a good idea.

If you don't want to directly add the notFound error to the list of errors returned to the user, then maybe you can embed it into the Flavor struct.

File: pkg/asset/installconfig/openstack/validation/machinepool.go
Comment (staebler) [inline pkg/asset/installconfig/openstack/validation/machinepool.go:107]: The note is not accurate. It is perfectly fine to unpack a nil slice.
```suggestion
func validateFlavor(flavorName string, ci *CloudInfo, req flavorRequirements, fldPath *field.Path) field.ErrorList {
	if flavorName == "" {
		return nil
	}
```

File: pkg/asset/installconfig/openstack/validation/cloudinfo.go
Comment (staebler) [inline pkg/asset/installconfig/openstack/validation/cloudinfo.go:184]:
```suggestion
	return errors.As(err, &errNotFound) || strings.Contains(err.Error(), "Resource not found")
```

File: pkg/asset/installconfig/openstack/validation/cloudinfo.go
Comment (staebler) [inline pkg/asset/installconfig/openstack/validation/cloudinfo.go:48]: It is not clear to me why the flavor is a special case that requires storing `userError`. For example, if the OS image is not found, then validation will yield a generic "NotFound" error. This is accomplished by checking that `OSImage` in the cloud info is nil. Can we not similarly check if `ci.Flavors[flavorName].Flavor == nil` to determine if the flavor was not found?

File: pkg/asset/installconfig/openstack/validation/machinepool.go
Comment (iamemilio) [reply pkg/asset/installconfig/openstack/validation/machinepool.go:107]: ack, will fix

File: pkg/asset/installconfig/openstack/validation/machinepool.go
Comment (iamemilio) [reply pkg/asset/installconfig/openstack/validation/machinepool.go:111]: yeah the check for ok doesnt do anything actually

**Comment by iamemilio** [comment]
/retest 

I think I took everyones points into consideration for the latest draft. PTAL

File: pkg/asset/installconfig/openstack/validation/cloudinfo.go
Comment (iamemilio) [reply pkg/asset/installconfig/openstack/validation/cloudinfo.go]: didnt know that! will fix

**Review by Fedosin — APPROVED** [review]
/lgtm
/approve

File: pkg/asset/installconfig/openstack/validation/machinepool.go
Comment (pierreprinetti) [inline pkg/asset/installconfig/openstack/validation/machinepool.go:131]: nit: This can stay `nil` :)

**Review by pierreprinetti — APPROVED** [review]
I like how this last version of the patch makes it consistent with the other resource checks (without hurting my feelings with unnecessary additional pointers :) )

I will /lgtm again if you plan to squash. Thanks!

File: pkg/asset/installconfig/openstack/validation/machinepool.go
Comment (iamemilio) [reply pkg/asset/installconfig/openstack/validation/machinepool.go:131]: ack, ill fix it

**Comment by iamemilio** [comment]
/hold cancel

Squashed and ready for final review

**Pull request merged**
